Compile fails on Solaris/x64 with syntax error in assembler

I'm using Sun C v5.13 compiler and the configure line:

```
$ Configure --prefix=/export/home/nse/workspace/openssl-solaris10_x64_sunc513/install shared enable-trace -m64 solaris64-x86_64-cc
```
Which produced this configure output:

```
Command line (with current working directory = .):

    /export/home/nse/workspace/openssl-solaris10_x64_sunc513/local_perl/bin/perl ../Configure --prefix=/export/home/nse/workspace/openssl-solaris10_x64_sunc513/install shared enable-trace -m64 solaris64-x86_64-cc

Perl information:

    /export/home/nse/workspace/openssl-solaris10_x64_sunc513/local_perl/bin/perl
    5.24.0 for i86pc-solaris

Enabled features:

    aria
    asm
    async
    autoalginit
    autoerrinit
    autoload-config
    bf
    blake2
    camellia
    capieng
    cast
    chacha
    cmac
    cmp
    cms
    comp
    ct
    deprecated
    des
    dgram
    dh
    dsa
    dso
    dtls
    dynamic-engine
    ec
    ec2m
    ecdh
    ecdsa
    engine
    err
    filenames
    fips
    gost
    idea
    legacy
    md4
    mdc2
    module
    multiblock
    nextprotoneg
    pinshared
    ocb
    ocsp
    padlockeng
    pic
    poly1305
    posix-io
    psk
    rc2
    rc4
    rdrand
    rfc3779
    rmd160
    scrypt
    secure-memory
    seed
    shared
    siphash
    siv
    sm2
    sm3
    sm4
    sock
    srp
    srtp
    sse2
    ssl
    static-engine
    stdio
    tests
    threads
    tls
    trace
    ts
    ui-console
    whirlpool
    tls1
    tls1-method
    tls1_1
    tls1_1-method
    tls1_2
    tls1_2-method
    tls1_3
    dtls1
    dtls1-method
    dtls1_2
    dtls1_2-method

Disabled features:

    afalgeng            [not-linux]      OPENSSL_NO_AFALGENG
    asan                [default]        OPENSSL_NO_ASAN
    buildtest-c++       [default]        
    crypto-mdebug       [default]        OPENSSL_NO_CRYPTO_MDEBUG
    devcryptoeng        [default]        OPENSSL_NO_DEVCRYPTOENG
    ec_nistp_64_gcc_128 [default]        OPENSSL_NO_EC_NISTP_64_GCC_128
    egd                 [default]        OPENSSL_NO_EGD
    external-tests      [default]        OPENSSL_NO_EXTERNAL_TESTS
    fuzz-libfuzzer      [default]        OPENSSL_NO_FUZZ_LIBFUZZER
    fuzz-afl            [default]        OPENSSL_NO_FUZZ_AFL
    ktls                [default]        OPENSSL_NO_KTLS
    makedepend          [unavailable]    
    md2                 [default]        OPENSSL_NO_MD2 (skip crypto/md2)
    msan                [default]        OPENSSL_NO_MSAN
    rc5                 [default]        OPENSSL_NO_RC5 (skip crypto/rc5)
    sctp                [default]        OPENSSL_NO_SCTP
    ssl-trace           [default]        OPENSSL_NO_SSL_TRACE
    ubsan               [default]        OPENSSL_NO_UBSAN
    unit-test           [default]        OPENSSL_NO_UNIT_TEST
    uplink              [no uplink_arch] OPENSSL_NO_UPLINK
    weak-ssl-ciphers    [default]        OPENSSL_NO_WEAK_SSL_CIPHERS
    zlib                [default]        
    zlib-dynamic        [default]        
    ssl3                [default]        OPENSSL_NO_SSL3
    ssl3-method         [default]        OPENSSL_NO_SSL3_METHOD

Config target attributes:

    AR => "ar",
    ARFLAGS => "r",
    CC => "cc",
    CFLAGS => "-xO5 -xdepend -xbuiltin",
    HASHBANGPERL => "/usr/bin/env perl",
    RANLIB => "ranlib",
    RC => "windres",
    asm_arch => "x86_64",
    bn_ops => "SIXTY_FOUR_BIT_LONG",
    build_file => "Makefile",
    build_scheme => [ "unified", "unix" ],
    cflags => "-xarch=generic64 -xstrconst -Xa",
    cppflags => "-D_REENTRANT",
    defines => [  ],
    disable => [  ],
    dso_scheme => "dlfcn",
    enable => [  ],
    ex_libs => "-lsocket -lnsl -ldl -lpthread",
    includes => [  ],
    lflags => "-mt",
    lib_cflags => "",
    lib_cppflags => "-DFILIO_H -DL_ENDIAN",
    lib_defines => [ "OPENSSL_BUILDING_OPENSSL" ],
    module_cflags => "-KPIC",
    module_cxxflags => undef,
    module_ldflags => "-G -dy -z text -Wl,-Bsymbolic",
    multilib => "/64",
    perl_platform => "Unix",
    perlasm_scheme => "elf",
    shared_cflag => "-KPIC",
    shared_defflag => "-Wl,-M,",
    shared_defines => [  ],
    shared_ldflag => "-G -dy -z text -Wl,-Bsymbolic",
    shared_rcflag => "",
    shared_sonameflag => "-Wl,-h,",
    shared_target => "solaris",
    thread_defines => [  ],
    thread_scheme => "pthreads",
    unistd => "<unistd.h>",

Recorded environment:

    AR = 
    BUILDFILE = 
    CC = 
    CFLAGS = 
    CPPFLAGS = 
    CROSS_COMPILE = 
    CXX = 
    CXXFLAGS = 
    HASHBANGPERL = 
    LDFLAGS = 
    LDLIBS = 
    OPENSSL_LOCAL_CONFIG_DIR = 
    PERL = 
    RANLIB = 
    RC = 
    RCFLAGS = 
    WINDRES = 
    __CNF_CFLAGS = 
    __CNF_CPPDEFINES = 
    __CNF_CPPFLAGS = 
    __CNF_CPPINCLUDES = 
    __CNF_CXXFLAGS = 
    __CNF_LDFLAGS = 
    __CNF_LDLIBS = 

Makevars:

    AR              = ar
    ARFLAGS         = r
    CC              = cc
    CFLAGS          = -xO5 -xdepend -xbuiltin -m64
    CPPDEFINES      = 
    CPPFLAGS        = 
    CPPINCLUDES     = 
    CXXFLAGS        = -m64
    HASHBANGPERL    = /usr/bin/env perl
    LDFLAGS         = 
    LDLIBS          = 
    PERL            = /export/home/nse/workspace/openssl-solaris10_x64_sunc513/local_perl/bin/perl
    RANLIB          = ranlib
    RC              = windres
    RCFLAGS         = 

NOTE: These variables only represent the configuration view.  The build file
template may have processed these variables further, please have a look at the
build file for more exact data:
    Makefile

build file:

    Makefile

build file templates:

    ../Configurations/common0.tmpl
    ../Configurations/unix-Makefile.tmpl
    ../Configurations/common.tmpl
```

During the build the error happens here:

```
...
cc -KPIC -xarch=generic64 -xstrconst -Xa -xO5 -xdepend -xbuiltin -m64 -DFILIO_H -DL_ENDIAN -DOPENSSL_BUILDING_OPENSSL -DOPENSSL_PIC -DOPENSSLDIR="\"/export/home/nse/workspace/openssl-solaris10_x64_sunc513/install/ssl\"" -DENGINESDIR="\"/export/home/nse/workspace/openssl-solaris10_x64_sunc513/install/lib/engines-3\"" -DMODULESDIR="\"/export/home/nse/workspace/openssl-solaris10_x64_sunc513/install/lib/ossl-modules\"" -D_REENTRANT -DNDEBUG  -c -o crypto/aes/libcrypto-lib-aes-x86_64.o crypto/aes/aes-x86_64.s
cc: Warning: -xarch=generic64 is deprecated, use -m64 to create 64-bit programs
Assembler: 
	"crypto/aes/aes-x86_64.s", line 2659 : Syntax error
	Near line: "	.section ".note.gnu.property", "a""
	"crypto/aes/aes-x86_64.s", line 2665 : Illegal mnemonic
	Near line: "	.asciz "GNU""
	"crypto/aes/aes-x86_64.s", line 2665 : Syntax error
	Near line: "	.asciz "GNU""
	"crypto/aes/aes-x86_64.s", line 2665 : Syntax error
	Near line: "	.asciz "GNU""
cc: fbe failed for crypto/aes/aes-x86_64.s
*** Error code 2
make: Fatal error: Command failed for target `crypto/aes/libcrypto-lib-aes-x86_64.o'
Current working directory /export/home/nse/workspace/openssl-solaris10_x64_sunc513/build-solaris10_x64_sunc513
```

> Ah! From [a manual I found](https://ftp.gnu.org/old-gnu/Manuals/gas-2.9.1/html_chapter/as_7.html):
> 
> > For ELF targets, the assembler supports another type of .section directive for compatibility with the Solaris assembler:
> > .section "name"[, flags...]
> > Note that the section name is quoted. There may be a sequence of comma separated flags:
> > #alloc
> > section is allocatable
> > #write
> > section is writable
> > #execinstr
> > section is executable
> 
> So maybe works more universally?
> 
> ```
> 	.section ".note.gnu.property", #alloc
> ```

This doesn't work on Linux:

crypto/aes/aes-x86_64.s: Assembler messages:
crypto/aes/aes-x86_64.s:2659: Error: character following name is not '#'
make[1]: *** [Makefile:18608: crypto/aes/libfips-lib-aes-x86_64.o] Error 1

--- a/crypto/perlasm/x86_64-xlate.pl
+++ b/crypto/perlasm/x86_64-xlate.pl
@@ -83,6 +83,10 @@ my $PTR=" PTR";
 my $nasmref=2.03;
 my $nasm=0;
 
+# GNU as indicator, as opposed to $gas, which indicates acceptable
+# syntax
+my $gnuas=0;
+
 if    ($flavour eq "mingw64")	{ $gas=1; $elf=0; $win64=1;
 				  $prefix=`echo __USER_LABEL_PREFIX__ | $ENV{CC} -E -P -`;
 				  $prefix =~ s|\R$||; # Better chomp
@@ -100,6 +104,12 @@ elsif (!$gas)
     $elf=0;
     $decor="\$L\$";
 }
+# Find out if we're using GNU as
+elsif (`$ENV{CC} -Wa,-v -c -o /dev/null -x assembler /dev/null 2>&1`
+		=~ /GNU assembler version ([2-9]\.[0-9]+)/)
+{
+    $gnuas=1;
+}
 
 my $cet_property;
 if ($flavour =~ /elf/) {
@@ -108,8 +118,10 @@ if ($flavour =~ /elf/) {
 	# with Intel CET support in order for linker to mark output with
 	# Intel CET support.
 	my $p2align=3; $p2align=2 if ($flavour eq "elf32");
+	my $section='.note.gnu.property, #alloc';
+	$section='".note.gnu.property", "a"' if $gnuas;
 	$cet_property = <<_____;
-	.section ".note.gnu.property", "a"
+	.section $section
 	.p2align $p2align
 	.long 1f - 0f
 	.long 4f - 1f

